enhancement(http source): URL path configuration flexibility, path also exposed as a new field

prognant opened the pull request:
Closes #5112 

Add three options to the http source:
* `url_path`: the path where log event should be posted (default: /)
* `path_key`: the name of the key used to export the path for each log even (default: path)
* `strict_path`: whether strict path should be enforce, if true only event send to the exact url_path will be accepted, if false log sent to a path that start with the value of url_path will be accepted

For example:
```
url_path: "/events"
strict_path: false
```

Will accept events posted to "/events/1234" and "/events/5678".

Outside the http source there is no behaviour change.

Open question: should we accept all URLs for the `heroku_logs` source as it would allow a user to send logs from multiple app to the same source.

jszwedko commented on src/sources/http.rs:
```diff
@@ -20,41 +17,47 @@ use tokio_util::codec::Decoder;
 use warp::http::{HeaderMap, HeaderValue, StatusCode};
 
 #[derive(Deserialize, Serialize, Debug, Clone)]
+#[serde(default)]
```
I think this (plus the `Default` trait implementation below) will have the undesired effect of not requiring the user to specify the `address` field in their config. We currently require this field.

On the other hand, I could see trying to choose a sensible default for it like you did here. I might suggest something other than `:80` as that is likely to be blocked.

For now, I'd recommend just replacing the `#[serde(default)]`s below and removing it from the struct. You can provide a default value for the field with something like:

https://github.com/timberio/vector/blob/c79545c74bb981c66e1082c12d635240c777aefd/src/sources/syslog.rs#L40-L41

https://github.com/timberio/vector/blob/c79545c74bb981c66e1082c12d635240c777aefd/src/sources/syslog.rs#L64-L66

prognant replied:
I think not changing the default behaviour should be favoured when possible, I followed your suggestion to keep the current mandatory behaviour for the `address` setting and rely on function to provided other default values.

jszwedko commented on docs/reference/components/sources/http.cue:
```diff
@@ -107,8 +107,38 @@ components: sources: http: {
 				}
 			}
 		}
+		url_path: {
```
What would you think of just calling this `path`? I think, in the context it appears, it'll be pretty clear it refers to an HTTP URL path.

jszwedko reviewed (commented):
Nice work @prognant !

I left one comment about configuration, but otherwise I think this looks good. I appreciate the comprehensive tests as well as support for `url_path` and `strict_path` config options.

In my opinion, we should expand this support to `heroku_logs` and any other HTTP-based sources for consistency. I think this would include: `prometheus_remote_write`, `splunk_hec`, `aws_kinesis_firehose`. We can split that off into a separate issue though.

prognant commented:
Thanks for the review, I addressed your comments  and created a follow-up issue to leverage this change in other relevant sources.

> Nice work @prognant !
> 
> I left one comment about configuration, but otherwise I think this looks good. I appreciate the comprehensive tests as well as support for `url_path` and `strict_path` config options.
> 
> In my opinion, we should expand this support to `heroku_logs` and any other HTTP-based sources for consistency. I think this would include: `prometheus_remote_write`, `splunk_hec`, `aws_kinesis_firehose`. We can split that off into a separate issue though.

jszwedko commented on src/sources/http.rs:
```diff
@@ -594,6 +668,96 @@ mod tests {
+        .await;
+
+        assert_eq!(
+            405,
```
I realized that 405, "Method not allowed", is a bit of an odd error for this condition. I would have expected a 404.

This looks like a bug in warp: https://github.com/seanmonstar/warp/issues/77

I think you could work around this by, instead of relying on warp's error handling with `path::end`, just accepting all requests and then, in the handler itself, when `strict_match` is true, check that the path matches and return a 404 manually if not.

jszwedko reviewed (commented):
Thanks @prognant ! I noticed one more thing, but otherwise this looks good to me.

bruceg commented on docs/reference/components/sources/http.cue:
```diff
@@ -107,8 +107,38 @@ components: sources: http: {
 		}
+		path: {
+			common:      false
+			description: "The URL path on which log event POST request shall be sent."
```
I think "request" needs to be plural:
```suggestion
			description: "The URL path on which log event POST requests shall be sent."
```

bruceg commented on docs/reference/components/sources/http.cue:
```diff
@@ -107,8 +107,38 @@ components: sources: http: {
+		strict_path: {
+			common: false
+			description: """
+				If set to true only request using the exact URL path specified in path will be accepted,
+				else event log sent to a URL path that starts with the value of path will be accepted.
+				With strict_path set to false and path set to \"/\" the configured HTTP source will
+				accept log events from any URL path.
```
We have also preferred to back-quote literals and config fields:
```suggestion
				If set to `true`, only requests using the exact URL path specified in `path` will be accepted;
				otherwise requests sent to a URL path that starts with the value of `path` will be accepted.
				With `strict_path` set to `false` and `path` set to `""`, the configured HTTP source will
				accept requests from any URL path.
```

bruceg commented on docs/reference/components/sources/http.cue:
```diff
@@ -107,8 +107,38 @@ components: sources: http: {
+		}
+		path_key: {
+			common:      false
+			description: "The event key in which the requested URL path used to send the log event will be stored."
```
```suggestion
			description: "The event key in which the requested URL path used to send the request will be stored."
```

bruceg commented on docs/reference/components/sources/http.cue:
```diff
@@ -122,6 +152,14 @@ components: sources: http: {
 					}
 				}
+				path: {
+					description: "The HTTP path the event was received from. The key can be changed using the path_key configuration setting"
```
```suggestion
					description: "The HTTP path the event was received from. The key can be changed using the `path_key` configuration setting"
```

bruceg commented on docs/reference/components/sources/http.cue:
```diff
@@ -135,13 +173,22 @@ components: sources: http: {
 					type: "*": {}
 				}
+				path: {
+					description: "The HTTP path the event was received from. The key can be changed using the path_key configuration setting"
```
```suggestion
					description: "The HTTP path the event was received from. The key can be changed using the `path_key` configuration setting"
```

bruceg commented on src/sources/http.rs:
```diff
@@ -119,6 +148,16 @@ impl SourceConfig for SimpleHttpConfig {
+    for event in events.iter_mut() {
+        event
+            .as_mut_log()
+            .insert(key, Value::from(path.to_string()));
```
FWIW this should be equivalent to:
```suggestion
            .insert(key, path.to_string().into());
```
(Not really a suggestion to change it, just a note)

bruceg reviewed (commented):
The code looks good, except for the HTTP response code issue identified by @jszwedko. Otherwise, I just have a few wording suggestions for the docs.

jszwedko commented on src/sources/util/http.rs:
```diff
@@ -202,11 +202,19 @@ pub trait HttpSource: Clone + Send + Sync + 'static {
+                        debug!(message = "Path rejected.");
+                        Err(warp::reject::custom(
+                            ErrorMessage::new(StatusCode::NOT_FOUND,
+                            StatusCode::NOT_FOUND.canonical_reason().unwrap().to_string())
```
Could we use `unwrap_or_default()` here? I realize that this particular code always has a canonical reason, but `unwrap()`s tend to jump out.

prognant replied:
That sounds safer. In this case I think we may directly go for `"Not found".to_string()` WDYT ?

jszwedko replied:
I'm also fine with that 😄

jszwedko replied:
It looks like there is also a `warp::reject::not_found()` you could use instead of `warp::reject::custom()`:

https://docs.rs/warp/0.2.5/warp/reject/fn.not_found.html

jszwedko reviewed (approved):
Nice work!

binarylogic commented:
Nice work!
